### DISCUSSION TITLE
AsyncSession doesn't support legacy Query method ?

### DISCUSSION THREAD

**Discussion by graille** [sqlalchemy/sqlalchemy#7971] [Usage Questions]
Hello everyone,

I'm using SQLAlchemy 1.4.35
With the implementation of the [beta AsyncIO interface for SQLAlchemy](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html), I'm trying to move my entire old-fashioned-non-optimized code, which used static engine and Session, to the new async engine and AsyncSession features in order to improve code readability and performances.

This is the kind of code I used to make before. It uses `asyncio.Lock` locks to prevent asynchronous errors when using sessions:

```python
async def update_channel_statistics(self):
    # Get users "User" instances from database
    await self.lock.acquire()
    with Session.begin() as current_session:
        current_stream, stream_live_data = self.core.get_current_twitch_stream()

        # Get last statistics registered for the current stream
        last_instant_stats: Optional[StreamInstantStats] = current_session \
            .query(StreamInstantStats) \
            .filter(StreamInstantStats.stream == current_stream) \
            .order_by(desc(StreamInstantStats.current_time)) \
            .limit(1) \
            .one_or_none()
        # Stuff...
   self.lock.release()
```
(I also used to put a try/finally to ensure `self.lock` was released).

Now, this is new code I'm trying to implement using AsyncSession :

```python
async def update_channel_statistics(self):
    # Get users "User" instances from database
    async with async_session.begin() as current_session:
        current_stream, stream_live_data = await current_session.run_sync(self.core.get_current_twitch_stream)

        # Get last statistics registered for the current stream
        last_instant_stats: Optional[StreamInstantStats] = await current_session \
            .query(StreamInstantStats) \
            .filter(StreamInstantStats.stream == current_stream) \
            .order_by(desc(StreamInstantStats.current_time)) \
            .limit(1) \
            .one_or_none()
        # Stuff...
```

But here is the issue. 
It seems that my IDE doesn't recognize query as an method of AsyncSession. 

According to the documentation, it seems that AsyncSession only support 2.0 style, I'm right ?
Any plan to integrate the legacy Query method with AsyncSession ? It's really strange since `current_session.run_sync` pass a Session to the called function, and authorize the use of legacy query method, but not AsyncSession...

**Comment by CaselIT** [comment]
Hi,

That's the expected behaviour, and there is currently no plan on implementing query in the async session.

**Reply by lschnoller** [reply]
does anyone know why?

**Reply by CaselIT** [reply]
It was decided to empatize the new select api, implementing new features in it

**Reply by zzzeek** [reply]
the Query class uses synchronous methods so it can't be used as is within async

Query makes it unclear which methods are actually doing IO or not, so would be tedious to use in an async context
